### PR TITLE
Manually parse cookies

### LINKED ISSUE #3778: [crunchyroll] Endless loop when trying to download an episode
### Checklist

- [X] I'm reporting a broken site
- [X] I've verified that I'm running yt-dlp version **2022.05.18** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/ytdl-org/youtube-dl#video-url-contains-an-ampersand-and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [ ] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

US

### Description

I can watch episodes in the browser just fine, but trying to download something causes yt-dlp to enter an endless loop.

I used a Chrome addon to generate a cookies.txt for Crunchyroll, but yt-dlp appears to either ignore it or not use the data contained in it because it complains about not being logged in. I used to be able to download things before version 2022.05.18 using a cookies file, though I don't remember which version I was using before.

### Verbose log

```shell
$ yt-dlp --cookies cookies.txt -vU "https://beta.crunchyroll.com/watch/G4VUQEQV9/the-friendship-scheme"
[debug] Command-line config: ['--cookies', 'cookies.txt', '-vU', 'https://beta.crunchyroll.com/watch/G4VUQEQV9/the-friendship-scheme']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.05.18 [b14d52355]
[debug] Python version 3.10.4 (CPython 64bit) - macOS-12.3.1-arm64-arm-64bit
[debug] Checking exe version: ffprobe -bsfs
[debug] Checking exe version: ffmpeg -bsfs
[debug] exe versions: ffmpeg 5.0.1 (setts), ffprobe 5.0.1, rtmpdump 2.4
[debug] Optional libraries: Cryptodome-3.14.1, brotli-1.0.9, certifi-2021.10.08, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.3
[debug] Proxy map: {}
Latest version: 2022.05.18, Current version: 2022.05.18
yt-dlp is up to date (2022.05.18)
[debug] [crunchyroll:beta] Extracting URL: https://beta.crunchyroll.com/watch/G4VUQEQV9/the-friendship-scheme
[crunchyroll:beta] the-friendship-scheme: Downloading webpage
[crunchyroll:beta] the-friendship-scheme: Not logged in. Redirecting to non-beta site - https://www.crunchyroll.com/spy-x-family/the-friendship-scheme-843099
[debug] [crunchyroll] Extracting URL: https://www.crunchyroll.com/spy-x-family/the-friendship-scheme-843099
[crunchyroll] 843099: Downloading webpage
[crunchyroll] 843099: Redirected to beta site - https://beta.crunchyroll.com/watch/G4VUQEQV9/the-friendship-scheme
[debug] [crunchyroll:beta] Extracting URL: https://beta.crunchyroll.com/watch/G4VUQEQV9/the-friendship-scheme
[crunchyroll:beta] the-friendship-scheme: Downloading webpage
[crunchyroll:beta] the-friendship-scheme: Not logged in. Redirecting to non-beta site - https://www.crunchyroll.com/spy-x-family/the-friendship-scheme-843099
[debug] [crunchyroll] Extracting URL: https://www.crunchyroll.com/spy-x-family/the-friendship-scheme-843099
[crunchyroll] 843099: Downloading webpage
[crunchyroll] 843099: Redirected to beta site - https://beta.crunchyroll.com/watch/G4VUQEQV9/the-friendship-scheme
[debug] [crunchyroll:beta] Extracting URL: https://beta.crunchyroll.com/watch/G4VUQEQV9/the-friendship-scheme
[crunchyroll:beta] the-friendship-scheme: Downloading webpage
[crunchyroll:beta] the-friendship-scheme: Not logged in. Redirecting to non-beta site - https://www.crunchyroll.com/spy-x-family/the-friendship-scheme-843099
[debug] [crunchyroll] Extracting URL: https://www.crunchyroll.com/spy-x-family/the-friendship-scheme-843099
[crunchyroll] 843099: Downloading webpage
[crunchyroll] 843099: Redirected to beta site - https://beta.crunchyroll.com/watch/G4VUQEQV9/the-friendship-scheme
[debug] [crunchyroll:beta] Extracting URL: https://beta.crunchyroll.com/watch/G4VUQEQV9/the-friendship-scheme
[crunchyroll:beta] the-friendship-scheme: Downloading webpage
[crunchyroll:beta] the-friendship-scheme: Not logged in. Redirecting to non-beta site - https://www.crunchyroll.com/spy-x-family/the-friendship-scheme-843099
[debug] [crunchyroll] Extracting URL: https://www.crunchyroll.com/spy-x-family/the-friendship-scheme-843099
[crunchyroll] 843099: Downloading webpage
```

### LINKED ISSUE #4776: `InfoExtractor._get_cookies` fails if values contain quotes
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting a bug unrelated to a specific site
- [X] I've verified that I'm running yt-dlp version **2022.08.19** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/ytdl-org/youtube-dl#video-url-contains-an-ampersand-and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Provide a description that is worded well enough to be understood

`InfoExtractor._get_cookies` uses `http.cookies.SimpleCookie` to process the cookies. Analogue to #4692 the parsing will fail fast instead of skipping the invalid values.

SimpleCookie allows values with quotes if set explicitly.

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['--cookies-from-browser', 'firefox', '-j', 'https://beta.crunchyroll.com/de/watch/GG1U2Q50J/the-former-couple-refuses-to-say', '-vU']
[debug] Encodings: locale cp1252, fs utf-8, pref cp1252, out utf-8, error utf-8, screen utf-8 
[debug] yt-dlp version 2022.08.19 [48c88e0] (pip) 
[debug] Python 3.10.5 (CPython 64bit) - Windows-10-10.0.22000-SP0  
[debug] Checking exe version: ffmpeg -bsfs 
[debug] Checking exe version: ffprobe -bsfs 
[debug] exe versions: ffmpeg n5.1-10-g6ee1996721-20220822 (setts), ffprobe n5.1-10-g6ee1996721-20220822 
[debug] Optional libraries: Cryptodome-3.15.0, brotli-1.0.9, certifi-2022.06.15, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.3
[Cookies] Extracting cookies from firefox
[debug] Extracting cookies from: "C:\Users\grub4k\AppData\Roaming\Mozilla\Firefox\Profiles\kbiex092.default-release\cookies.sqlite"
[Cookies] Extracted 790 cookies from firefox
[debug] Proxy map: {} 
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest 
Latest version: 2022.08.19, Current version: 2022.08.19 
yt-dlp is up to date (2022.08.19) 
[debug] [crunchyroll:beta] Extracting URL: https://beta.crunchyroll.com/de/watch/GG1U2Q50J/the-former-couple-refuses-to-say
[crunchyroll:beta] the-former-couple-refuses-to-say: Downloading webpage 
ERROR: GG1U2Q50J: An extractor error has occurred. (caused by KeyError('byId')); please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
  File "C:\Users\grub4k\AppData\Local\pypoetry\Cache\virtualenvs\crunchyload-sk9Pq0JJ-py3.10\lib\site-packages\yt_dlp\extractor\common.py", line 666, in extract
    ie_result = self._real_extract(url)
  File "C:\Users\grub4k\AppData\Local\pypoetry\Cache\virtualenvs\crunchyload-sk9Pq0JJ-py3.10\lib\site-packages\yt_dlp\extractor\crunchyroll.py", line 805, in _real_extract
    return self._redirect_from_beta(url, lang, internal_id, display_id, True, CrunchyrollIE.ie_key())
  File "C:\Users\grub4k\AppData\Local\pypoetry\Cache\virtualenvs\crunchyload-sk9Pq0JJ-py3.10\lib\site-packages\yt_dlp\extractor\crunchyroll.py", line 752, in _redirect_from_beta
    content_data = initial_state['content']['byId'][internal_id]
KeyError: 'byId'
```

### PR DESCRIPTION
**Pull request by Grub4K**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

Iterate over split cookie chunks to skip invalid values instead of failing fast like `http.cookies.SimpleCookie` does. This allows to preserve all cookie values, even invalid ones.

Fixes #4776

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [x] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

File: yt_dlp/extractor/common.py [hunk 1]
```diff
@@ -3633,7 +3633,23 @@ def _set_cookie(self, domain, name, value, expire_time=None, port=None,
...
         """ Return a http.cookies.SimpleCookie with the cookies for the url """
-        return http.cookies.SimpleCookie(self._downloader._calc_cookies(url))
+        raw_cookie = self._downloader._calc_cookies(url)
+
+        cookie = http.cookies.SimpleCookie()
+        if raw_cookie is None:
+            return cookie
+
+        # Iterate over split chunks to skip invalid values instead of failing fast.
+        for chunk in raw_cookie.split(";"):
```
Comment (dirkf) [inline yt_dlp/extractor/common.py:3643]: `;` is a valid character in a `quoted-string` cookie value.
```
>>> http.cookies.SimpleCookie('''chips="a;hoy"; vienna=finger''')
<SimpleCookie: chips='a;hoy' vienna='finger'>
>>> 
```
Check https://github.com/python/cpython/blob/3.10/Lib/http/cookies.py for what is currently being done (and has been for 22 years, according to the header comments).

What could additionally be done is to add a parallel pattern to the value match in `_CookiePattern`, s/t like this
```
r"""
(?:
  (?P<val>                         # Start of group 'val'
      "(?:[^\\"]|\\.)*"                  # Any doublequoted string
      |                                  # or
      \w{3},\s[\w\d\s-]{9,11}\s[\d:]{8}\sGMT  # Special case for "expires" attr
      |                                  # or
      [""" + _LegalValueChars + r"""]*      # Any word or empty string
      )|                                # End of group 'val'
  (?P<bad>(?:\\;|[^;])*)           # fallback match to skip bad value 
)
"""
Then in a rewrite of the `BaseCookie.__parse_string()` method, add a step to check the `bad` group: if it matches, skip past the match and try to parse the cookie string from there.

Comment (coletdjnz) [reply yt_dlp/extractor/common.py:3643]: Yeah cookie parsing is fairly complicated and needs to obey the relevant RFCs, so it's not going to be simple.

**Comment by coletdjnz** [comment]
Relevant CPython issue: https://github.com/python/cpython/issues/92936

File: yt_dlp/extractor/common.py (hunk 1 above)
Comment (coletdjnz) [reply yt_dlp/extractor/common.py:3643]: Though this does look to be what django does https://github.com/django/django/commit/93a135d111c2569d88d65a3f4ad9e6d9ad291452

Comment (Grub4K) [reply yt_dlp/extractor/common.py:3643]: RFC2109 has been obsoleted by RFC6265 and suggests semicolons are in fact not valid as can be read in [RFC6265, Section 4.1.1](https://www.rfc-editor.org/rfc/rfc6265#section-4.1.1) about `cookie-octet` and `cookie-name` (which forwards to [RFC2616, Section 2.2, `token`](https://www.rfc-editor.org/rfc/rfc2616#section-2.2)).

Comment (pukkandan) [reply yt_dlp/extractor/common.py:3643]: Even so, follow CPython as closely as possible, making only minimum required change to fix our issue

Comment (Grub4K) [reply yt_dlp/extractor/common.py:3643]: In my opinion the current way of parsing the cookie string cannot easily be fixed to handle the strings as required.

This change requires subclassing `SimpleCookie` and completely reimplementing its `load()` or `__parse_string()` function with a custom regex and behavior. This new subclass will then be returned from `_get_cookies()` instead. Let me know if this is the desired behavior or, if not, how and where the implementation should be.

Comment (Grub4K) [reply yt_dlp/extractor/common.py]: I have implemented a wrapper function `__make_simple_cookie()` which closely resembles the parsing done in `http.cookies` (except for the `bad` group and not failing on errors). Please review anew.

**Review by dirkf — COMMENTED** [review]
That looks like what I had in mind, thanks. But the proof of the coding is in the testing!

Regarding RFC6265, it's possible that, as with the previous cookie RFCs, the real world doesn't match. I haven't studied the changes in the spec.

**Comment by Grub4K** [comment]
All right, I am composing a testing suite for the cookie parsing, derived from the CPython tests.
Will push the tests to `test/test_YoutubeDL.py` when they are done.

**Review by dirkf — COMMENTED** [review]
Supposing that the tests all pass this is nice.

Perhaps consider casting it as a separate class, which might eventually get moved out of `extractor/common.py`? According to the docs and CPython 3.10 source, s/t like this ought to work:
```
class ImprovedSimpleCookie(http.cookies.SimpleCookie):
    # your code from l.3639-3678, re-indented as necessary
    def load(self, data):
        if not isinstance(data, str):
            super().load(data)
            return
        # your code from l.3681-3724
...
    def _get_cookies(self, url):
        """ Return a SimpleCookie with the cookies for the url """
        return ImprovedSimpleCookie(self._downloader._calc_cookies(url))
```
Although it isn't used in yt-dl/dlp (I think), the not-str case is a dict of (cookie_name, value) which the original Cookie/http.cookies code (2.6+) should handle.

**Comment by Grub4K** [comment]
Extracting the class immediately seems to make more sense, potentially placing it into `cookies.py` (with the tests moved to `test_cookies.py`)? Or should the class reside within `common.py` for now?

**Comment by dirkf** [comment]
Yes, that file isn't in yt-dl but looks plausible: take advice from @pukkandan.

**Comment by pukkandan** [comment]
> Extracting the class immediately seems to make more sense, potentially placing it into `cookies.py` (with the tests moved to `test_cookies.py`)?

That makes sense

**Comment by dirkf** [comment]
Late now, but `Lenient` doesn't properly describe what is being done, ie fixing a defect in the parsing implemented in `SimpleCookie.load()`. Any of `Fixed`, `Better`, `Improved`, `Correct` would be more accurate ...

**Comment by Grub4K** [comment]
It is not a defect but a design choice as the comment in the code suggests, and lenient describes the behavior of the cookie accurately.

**Pull request merged**